Odata-docs/webapi-8/fundamentals/actions-functions.md

```diff
@@ -0,0 +1,423 @@
+
+## URI patterns for calling OData actions and functions
+In the examples below `odata` is the route prefix.
+
+### Bound function
+The function is after an entity, entityset, singleton. `Books` and `Graphs` are entity sets.
+
+```
+GET ~/odata/Books/mostRecent()
+GET ~/odata/Graphs/Default.GetShapeCount()
+GET ~/odata/Graphs/Default.GetShapeCount(shapeType=FunctionActionBlog.ShapeType’Circle’)
+```
+
+`GetShapeCount` function has an overload that accepts a `ShapeType` parameter.
+
+We can bind to a single entity.
+
+`GET ~/odata/Books(1)/isChildBook()`
+
+We can also bind to a singleton
+
+`GET ~/odata/Me/isMyCalendarBlocked()`
+
+### Unbound function
+Function is after the route prefix.
+
+```
+GET ~/odata/ReturnAllForKidsBooks
+GET ~/odata/GetSalesTaxRate(10)
+```
+`ReturnAllForKidsBooks` doesn't have a parameter, while `GetSalesTaxRate` accepts one integer parameter.
+
+If `GetSalesTaxRate` had more than one parameter, the call would be
+
+`GET ~/odata/GetSalesTaxRate(10, "US")`
+
+Parameter-less functions can work with or without parenthesis `()` after function names.
+This is how we configure parameter-less functions to work without parenthesis.
+
+```csharp
+builder.Services.AddControllers().AddOData(opt =>
+{
+    opt.AddRouteComponents("odata", EdmModel.GetEdmModel()).Count().OrderBy().Filter().Select().Expand();
+    opt.RouteOptions.EnableNonParenthesisForEmptyParameterFunction = true;
+});
+```
+
+### Bound Action
+Actions are invoked via `POST` requests therefore we can pass a payload in the body.
+
+```json
+POST ~/odata/Books('1')/Rate
+{
+    "rating": 7
+}
+```
+
+### Unbound action
+```json
+POST ~/odata/incrementBookYear
+{
+    "increment": 7,
+    "id": "1"
+}
+```
+
+## Practical examples
+We will add detailed examples on how to configure bound and unbound functions and actions.
+We will start by setting up the ASP.NET Core OData v8.x project.
+
+### Data Model classes
+```csharp
+public class Book
+{
+    public string ID { get; set; }
+    public string Title { get; set; }
+    public bool ForKids { get; set; }
+}
+
+public class BookRating
+{
+    public string ID { get; set; }
+    public int Rating { get; set; }
+    public string BookID { get; set; }
+}
+```
+
+### Controller
+```csharp
+public class BooksController : ODataController
+{
+    // Get ~/Books
+    [EnableQuery]
+    public IActionResult Get()
+    {
+        return Ok(DataSource.Instance.Books);
+    }
+}
+```
+
+### Build Edm Model
+```csharp
+private static IEdmModel GetEdmModel()
+{
+    var builder = new ODataConventionModelBuilder();
+    builder.EntitySet<Book>("books");
+    var model = builder.GetEdmModel();
+    return model;
+}
+```
+
+## Bound Function
+Bound functions can be bound to an entity type or a collection of entity type.
+In our example model above, the bound function can be bound to a single book or a collection of books.
+We will demonstrated how to add a function that is bound to a collection.
+
```

Comment:
I think for the sake of being exhaustive, we should have examples for binding both to collections and single entities, with and without parameters, with and without return type. The reasoning is that this should be the authoritative source of the truth, if it's not mentioned here, we shouldn't expect the user to know how to use it or to know whether it exists.

Reply:
That will be too many examples in one section considering each example has code in the model & controller and request & response examples.

Reply:
I get that. But where else can the user get this information? That user doesn't have to read all the examples, but if they need to do something or to find out something, it would be great if they can find it in the docs. In my opinion, having "too many examples" is a result of OData supporting many different options. I think it's worse to support things we don't demonstrate how to use.